pkg/secrets/secrets.go

```diff
@@ -0,0 +1,211 @@
+package secrets
+
+import (
+	"path/filepath"
+	"time"
+
+	"github.com/containers/storage/pkg/lockfile"
+	"github.com/containers/storage/pkg/stringid"
+	"github.com/pkg/errors"
+)
+
+// ErrInvalidPath indicates that the secrets path is invalid
+var ErrInvalidPath = errors.New("Invalid secrets path")
+
+// ErrNoSuchSecret indicates that the the secret does not exist
+var ErrNoSuchSecret = errors.New("no such secret")
+
+// ErrSecretNameInUse indicates that the secret name is already in use
+var ErrSecretNameInUse = errors.New("secret name in use")
+
+// ErrInvalidSecretName indicates that the secret name is invalid
+var ErrInvalidSecretName = errors.New("invalid secret name")
+
+// secretsFile is the name of the file that the secrets database will be stored in
+var secretsFile = "secrets.json"
+
+// SecretsManager holds information on handling the secret
+type SecretsManager struct {
+	// secretsPath is the path to the db file where secrets are stored
+	secretsDBPath string
+	// driver is the secrets data driver
+	driver SecretsDriver
```

Comment:
Raising it here to open a discussion:

In the current form, there is one secrets manager per driver.  If we want to support more than one secret driver in the future, we need to move that info down to each individual secret.

Reply:
So rather than having `driver` in the `secretsManager` struct, it would be in the `Secret` struct? and store would be `(SecretsManager).Store(name, data, driver)`, and delete would be

```
func (s *SecretsManager) Delete(){
     secret=lookupSecret(nameOrID)
     secret.driver.delete(nameorID)
}
```

Reply:
Yes, the driver would be associated with the secret instead of the manager.  That would allow to manage all secrets across all drivers.

`containers.conf` could include a mapping from `driver` (string) to `binary-path` (string). So `podman ... --secret-driver foo` would yield a look-up in `containers.conf` and check if `foo` exists, and then call the binary to use the driver.

@rhatdan @mheon WDYT?

Reply:
SGTM

Reply:
SGTM here too.

Reply:
Okay, I was able to restructure `driver` to be under secrets like this:

```golang
type Secret struct {
	Name string `json:"name"`
	ID string `json:"id"`
	Driver SecretsDriver
}
```

but ran into a problem:  json is untyped and so I don't know how to store the secretsDriver interface in the json file without the secrets pkg being agnostic to the driver type and to retrieve later. (the json file is a `map[string]Secret` aka id:Secret)

I saw this as a solution: https://stackoverflow.com/questions/35583735/unmarshaling-into-an-interface-and-then-performing-type-assertion
But by doing this I'd be putting driver stuff directly into the secrets pkg, which I'm not sure is the right move. Also, if the secret wasn't in memory, I would have to end up reconstructing the driver every time? Not entirely sure what's going on
If you have any pointers on this @vrothberg let me know

Reply:
The `Driver` field should be of type string.  For instance:
```go
newSecret := Secret {
  Name: "foo",
  ID: "abcde1235....",
  Driver: "file",
}
```

The secret manager has to "know" and understand what "file" is.  In our case, that would be the file driver.

In the future, we may want to support more drivers.  Then, we need to think about how we can map `Secret.Driver` to a real secrets driver, which may be another program.

Reply:
Right it would either be a path to an executable or a new one that we set.  For example
"env" to set environment variables with a secret, that would not get committed to an image.

Reply:
Okay, so , for sake of less confusion, when I use `Driver` I mean the string version, and SecretsDriver I mean the interface 
Driver would be of type string, and it would be either a string representation of a driver that we implemented, (ie the filedriver) or a path to an executable. That part makes sense to me.
I'm still struggling to wrap my head around translating a string to a SecretsDriver though. For example,  if `Driver:"file"`, we would also need to know other stuff like `dataFilePath` for the filedriver to function (and I assume we might need other metadata for future secrets drivers as well? I'm not entirely familiar) Does this mean we should also include a `DriverOptions map[string]string` field in the secrets struct?
Also, if we do translate a string Driver to SecretsDriver, does this mean that every time we call an operation on a secret, we would have to make a new driver? (this seems strange to me so I'm making sure) ie, 
```
func (s *SecretsManager) Delete(nameOrID string) (string, error) {
     secret := s.getSecret(nameOrID)
     driver := driverify(secret.Driver) // secret.Driver is the string, ie "file"
     driver.delete(nameOrID)
}

func driverify (driver string, metadata map[string]string){
     switch driver{
     case "file":
         return NewFileDriver(metadata[path])
     case "whatever other driver:
         return NewWhateverDriver(stuff)
     }
}
```

Also, if we use an executable as the driver, we would need a way to pass options to the executable, right? Would that go into something like the DriverOptions map? and then we could create a executableDriver set of interface functions like we did with the filedriver, and all it does is call the executable at the given path with the options?

Hopefully I'm understanding this correctly, I just want to make sure :sweat_smile:

Reply:
The translation from `Secret.Driver` to  `SecretsDriver` for now only needs to understand "file".  With "file", we have our built-in secrets driver that implements the driver interface.  The important thing is that the implementation is all in our code base.

If we want to support other external secret drivers in the future, we would need to add a new driver implementation.  This driver would conceptually look as follows:
```func (d *ExternalDriver) List() (..., error) {
    // Execute external program
    command.Execute(driverBinary, "list")
}
```
Here we have 1 implementation of the driver interface that can serve N external drivers.  Docker has a plugin mechanism where external drivers _must_ implement a protocol.

We should ignore external drivers for now.  The fact that we _can_ support that in the future is enough :)  A `Secret.Driver` string other than "file" in the future may be path to a binary, a path to unix socket, or just another name that we may translate in the containers.conf.   Lot's of options but nothing to worry about now.